pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java

```diff
@@ -333,94 +306,162 @@ public void reloadSegment(String segmentName, IndexLoadingConfig indexLoadingCon
+      return;
     }
-    File indexDir = downloadSegment(segmentName, zkMetadata);
-    addSegment(indexDir, indexLoadingConfig);
-    LOGGER.info("Downloaded and loaded segment: {} of table: {} with crc: {}", segmentName, _tableNameWithType,
-        zkMetadata.getCrc());
-  }
 
-  protected boolean allowDownload(String segmentName, SegmentZKMetadata zkMetadata) {
-    return true;
+    try {
+      Schema schema = ZKMetadataProvider.getTableSchema(_propertyStore, _tableNameWithType);
+      if (!ImmutableSegmentLoader.needPreprocess(segmentDirectory, indexLoadingConfig, schema)) {
+        // The loaded segment is still consistent with current table config or schema.
+        LOGGER.info("Segment: {} of table: {} is consistent with table config and schema", segmentName,
+            _tableNameWithType);
+        loadSegment(segmentDirectory, indexLoadingConfig, schema);
+        return;
+      }
+      // If any discrepancy is found, get the segment from tier backend, reprocess and load it.
+      LOGGER.info("Segment: {} of table: {} needs reprocess to reflect latest table config and schema", segmentName,
+          _tableNameWithType);
+      // Close the stale SegmentDirectory object before loading the newly processed segment.
+      closeSegmentDirectoryQuietly(segmentDirectory);
+      downloadTierSegmentAndProcess(segmentName, indexLoadingConfig, schema);
+    } catch (Exception e) {
+      closeSegmentDirectoryQuietly(segmentDirectory);
+      LOGGER.error("Failed to reprocess and load segment: {} of table: {}", segmentName, _tableNameWithType, e);
+      downloadRawSegmentAndProcess(segmentName, indexLoadingConfig, zkMetadata,
+          ZKMetadataProvider.getTableSchema(_propertyStore, _tableNameWithType));
+    }
   }
 
-  protected File downloadSegment(String segmentName, SegmentZKMetadata zkMetadata)
+  /**
+   * Get the segment from the configured tier backend, reprocess it with the latest table
+   * config and schema and then load it. Please note that the segment is from tier backend,
+   * not deep store, for incremental processing.
+   */
+  private void downloadTierSegmentAndProcess(String segmentName, IndexLoadingConfig indexLoadingConfig,
+      @Nullable Schema schema)
       throws Exception {
-    // TODO: may support download from peer servers for RealTime table.
-    return downloadSegmentFromDeepStore(segmentName, zkMetadata);
+    SegmentDirectoryLoaderContext loaderContext =
+        new SegmentDirectoryLoaderContext(indexLoadingConfig.getTableConfig(), indexLoadingConfig.getInstanceId(),
+            segmentName, indexLoadingConfig.getSegmentDirectoryConfigs());
+    SegmentDirectoryLoader segmentDirectoryLoader =
+        SegmentDirectoryLoaderRegistry.getSegmentDirectoryLoader(indexLoadingConfig.getSegmentDirectoryLoader());
+    File indexDir = getSegmentDataDir(segmentName);
+    segmentDirectoryLoader.download(indexDir, loaderContext);
```

Comment:
It seems to me that the segment directory loader should take care of downloading from tier backend or deepstore, depending on some setting, perhaps in the loadercontext.

I am yet to understand what the difference is between a tier backend segment and the "raw" segment, though. Let me know what I am missing here.

Reply:
Preprocessed segments (or tier backend segment) are put on server's storage backend (not using local disk here on purpose) to serve queries. SegmentDirectoryLoader interface knows how to load those preprocessed segments, essentially to initialize objects like SegmentDirectory and ColumnIndexDirectory to get ready for queries. Those interfaces abstract away the impl details and also where the preprocessed segments physically are.

It's on purpose not putting logic around deep store behind SegmentDirectoryLoader interface. As deep store is not designed as server's storage backend to hold preprocessed segments. Deep store keeps the raw segments, not preprocessed with latest table config and schema for queries.

Reply:
The download and upload methods are introduced to SegmentDirectoryLoader interface is because today the SegmentPreprocessor and pretty much all kinds of IndexCreators used by it requires the `File indexDir` as an input param. If they were able to work with SegmentDirectory interface instead of `File indexDir`, the download/upload methods wouldn't be needed in SegmentDirectoryLoader interface. 

Hope this adds more clarity on why downloading raw segment from deep store is not part of SegmentDirectoryLoader interface. As there can be a path forward to remove download/upload methods from SegmentDirectoryLoader, if segment preprocessing logic could work with SegmentDirectory interface instead of requiring File index explicitly.

Reply:
`the segment directory loader should take care of downloading from tier backend or deepstore, depending on some setting, perhaps in the loadercontext.` -  @klsince wdyt of this suggestion? Could we put download logic entirely behind the impl? Or is that hard to do because of the dependencies of downloadFromDeepstore?

Reply:
Thanks for the feedbacks, @mcvsubbu ! We proposed a new way to refactor the segment loading logic flow in this [short doc](https://docs.google.com/document/d/1oTyqqfkb4ep03alDojCokBrLdwIj7lsqF8Nwjifumsw/edit). Please review!

Reply:
The doc does not convey any design details. It is simply a commentary on this PR. Please, include the text of the doc right here in the PR as a comment.

It is not clear (to me) what this PR is achieving. Is this a PR to just improve code readability? (you say "confusing" in your doc). Or, is this a PR to make Pinot more extensible for some purpose?

Reply:
This is to make Pinot more extensible and future proofing for supporting cases like lazy loading and tiering. Improvement of readability is being done simply to make the review go easier, since that part of code was a lil complex to read to begin with..

Reply:
Refactored with new way proposed above. Please review the updated code. Thanks!